bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManagerImpl.java

```diff
@@ -1033,6 +1043,32 @@ private void setThingStatus(Thing thing, ThingStatusInfo thingStatusInfo) {
         }
     }
 
+    @Override
+    public void setDisabled(Thing thing, boolean disabled) {
+        ThingHandler handler = thing.getHandler();
+
+        if (disabled) {
+            logger.debug("Thing {} will be disabled.", thing.getUID().getAsString());
+            if (handler != null) {
+                handler.dispose();
+            }
+            setThingStatus(thing, buildStatusInfo(ThingStatus.UNINITIALIZED, ThingStatusDetail.DISABLED));
+        } else {
+            handler.initialize();
+        }
+
+        // Persist the status of the updated thing.
+        managedThingStatusInfoProvider.update(new ManagedThingStatusInfo(thing.getUID(), thing.getStatusInfo()));
```

Comment:
because dispose and initialize might have async implementations we cannot easily persist the status here... additionally we cannot set the DISABLED status after disposing the handler

Reply:
Right, not the thing status - but the information that it should be disabled.

Reply:
It seems too early to persist the status indeed. As the disposal is delegated to the handler, we are not sure what will happen in each dispose() method and when the async disposal will finish. Anyway, the `dispose()` method of the handler might be an empty one, therefore with the current implementation there is no way to know if the dispose is finished or not. 
@sjka , @kaikreuzer what do you think about providing a defaoult implementaion or some sort of a callback in `BaseThingHandler`? If we want to be certain that the disposal is finished we havae to provide this information somehow.

Reply:
The `ThingHandler.dispose()` method is called synchronously from the framework. Therefore you can safely persist the boolean flag that this thing should be disabled. Then, later on in `doUnregisterHandler(...)` you can use this information to set the thing to `UNINITIALIZED/DISABLED` instead of `UNINITIALIZED/HANDLER_MISSING_ERROR`.